File: js/packages/teams-ai/src/embeddings/LLMClient.ts
import { ConversationHistory, FunctionRegistry, GPT3Tokenizer, Message, Prompt, PromptFunctions, PromptMemory, PromptSection, Tokenizer, Utilities, VolatileMemory } from "promptrix";
import { PromptResponse, Validation, PromptResponseValidator, PromptCompletionModel } from "./ai/types";
import { DefaultResponseValidator } from "./DefaultResponseValidator";
import { MemoryFork } from "./MemoryFork";
import { Colorize } from "./internals";
import { OpenAIModel } from "../models/OpenAIModel";
import { FunctionResponseValidator } from "./FunctionResponseValidator";

/**
 * Options for an LLMClient instance.
 */
export interface LLMClientOptions {
    /**
     * AI model to use for completing prompts.
     */
    model: PromptCompletionModel;

    /**
     * Prompt to use for the conversation.
     */
    prompt: PromptSection;

    /**
     * Optional. FunctionRegistry to use when expanding prompt template sections.
     * @remarks
     * An empty `FunctionRegistry` instance will be created by default
     */
    functions?: PromptFunctions;

    /**
     * Optional. Memory variable used for storing conversation history.
     * @remarks
     * The history will be stored as a `Message[]` and the variable defaults to `conversation`.
     */
    history_variable?: string;

    /**
     * Optional. Memory variable used for storing the users input message.
     * @remarks
     * The users input is expected to be a `string` but it's optional and defaults to `input`.
     */
    input_variable?: string;

    /**
     * Optional. Maximum number of conversation history messages to maintain.
     * @remarks
     * The number of tokens worth of history included in the prompt is controlled by the
     * `ConversationHistory` section of the prompt. This controls the automatic pruning of the
     * conversation history that's done by the LLMClient instance. This helps keep your memory from
     * getting too big and defaults to a value of `10` (or 5 turns.)
     */
    max_history_messages?: number;

    /**
     * Optional. Maximum number of automatic repair attempts the LLMClient instance will make.
     * @remarks
     * This defaults to a value of `3` and can be set to `0` if you wish to disable repairing of bad responses.
     */
    max_repair_attempts?: number;

    /**
     * Optional. Memory the LLMClient instance should use for rendering the prompt and persisting conversation history.
     * @remarks
     * If not specified a new instance of `VolatileMemory` will be created. Keep in mind that by using
     * VolatileMemory, any conversation history will be deleted when the LLMClient instance is deleted.
     */
    memory?: PromptMemory;

    /**
     * Optional. Tokenizer to use when rendering the prompt or counting tokens.
     * @remarks
     * If not specified a new instance of `GPT3Tokenizer` will be created.
     */
    tokenizer?: Tokenizer;

    /**
     * Optional. Response validator to use when completing prompts.
     * @remarks
     * If not specified a new instance of `DefaultResponseValidator` will be created. The
     * DefaultResponseValidator returns a `Validation` that says all responses are valid.
     */
    validator?: PromptResponseValidator;

    /**
     * Optional. If true, any repair attempts will be logged to the console.
     */
    logRepairs?: boolean;
}

/**
 * The configuration of the LLMClient instance.
 */
export interface ConfiguredLLMClientOptions {
    /**
     * AI model used for completing prompts.
     */
    model: PromptCompletionModel;

    /**
     * Memory variable used for storing conversation history.
     */
    history_variable: string;

    /**
     * Memory variable used for storing the users input message.
     */
    input_variable: string;

    /**
     *  FunctionRegistry used when expanding prompt template sections.
     */
    functions: PromptFunctions;

    /**
     * Maximum number of conversation history messages that will be persisted to memory.
     */
    max_history_messages: number;

    /**
     * Maximum number of automatic repair attempts that will be made.
     */
    max_repair_attempts: number;

    /**
     * Memory used for rendering the prompt and persisting conversation history.
     */
    memory: PromptMemory;

    /**
     * Prompt used for the conversation.
     */
    prompt: PromptSection;

    /**
     * Tokenizer used when rendering the prompt or counting tokens.
     */
    tokenizer: Tokenizer;

    /**
     * Response validator used when completing prompts.
     */
    validator: PromptResponseValidator;

    /**
     * If true, any repair attempts will be logged to the console.
     */
    logRepairs: boolean;
}

/**
 * LLMClient class that's used to complete prompts.
 *
 * @remarks
 * Each wave, at a minimum needs to be configured with a `client`, `prompt`, and `prompt_options`.
 *
 * Configuring the wave to use a `validator` is optional but recommended. The primary benefit to
 * using LLMClient is it's response validation and automatic response repair features. The
 * validator acts as guard and guarantees that you never get an malformed response back from the
 * model. At least not without it being flagged as an `invalid_response`.
 *
 * Using the `JSONResponseValidator`, for example, guarantees that you only ever get a valid
 * object back from `completePrompt()`. In fact, you'll get back a fully parsed object and any
 * additional response text from the model will be dropped. If you give the `JSONResponseValidator`
 * a JSON Schema, you will get back a strongly typed and validated instance of an object in
 * the returned `response.message.content`.
 *
 * When a validator detects a bad response from the model, it gives the model "feedback" as to the
 * problem it detected with its response and more importantly an instruction that tells the model
 * how it should repair the problem. This puts the wave into a special repair mode where it first
 * forks the memory for the conversation and then has a side conversation with the model in an
 * effort to get it to repair its response. By forking the conversation, this isolates the bad
 * response and prevents it from contaminating the main conversation history. If the response can
 * be repaired, the wave will un-fork the memory and use the repaired response in place of the
 * original bad response. To the model it's as if it never made a mistake which is important for
 * future turns with the model. If the response can't be repaired, a response status of
 * `invalid_response` will be returned.
 *
 * When using a well designed validator, like the `JSONResponseValidator`, the wave can typically
 * repair a bad response in a single additional model call. Sometimes it takes a couple of calls
 * to effect a repair and occasionally it won't be able to repair it at all. If your prompt is
 * well designed and you only occasionally see failed repair attempts, I'd recommend just calling
 * the wave a second time. Given the stochastic nature of these models, there's a decent chance
 * it won't make the same mistake on the second call. A well designed prompt coupled with a well
 * designed validator should get the reliability of calling these models somewhere close to 99%
 * reliable.
 *
 * This "feedback" technique works with all the GPT-3 generation of models and I've tested it with
 * `text-davinci-003`, `gpt-3.5-turbo`, and `gpt-4`. There's a good chance it will work with other
 * open source models like `LLaMA` and Googles `Bard` but I have yet to test it with those models.
 *
 * LLMClient supports OpenAI's functions feature and can validate the models response against the
 * schema for the supported functions. When an LLMClient is configured with both a `OpenAIModel`
 * and a `FunctionResponseValidator`, the model will be cloned and configured to send the
 * validators configured list of functions with the request. There's no need to separately
 * configure the models `functions` list, but if you do, the models functions list will be sent
 * instead.
 */
export class LLMClient extends (EventEmitter as { new(): LLMClientEmitter }) {
    /**
     * Configured options for this LLMClient instance.
     */
    public readonly options: ConfiguredLLMClientOptions;

    /**
     * Creates a new `LLMClient` instance.
     * @param options Options to configure the instance with.
     */
    public constructor(options: LLMClientOptions) {
        super();
        this.options = Object.assign({
            functions: new FunctionRegistry(),
            history_variable: 'history',
            input_variable: 'input',
            max_history_messages: 10,
            max_repair_attempts: 3,
            memory: new VolatileMemory(),
            tokenizer: new GPT3Tokenizer(),
            logRepairs: false
        }, options) as ConfiguredLLMClientOptions;

        // Create validator to use
        if (!this.options.validator) {
            // Check for an OpenAI model using functions
            if (this.options.model instanceof OpenAIModel && this.options.model.options.functions) {
                this.options.validator = new FunctionResponseValidator(this.options.model.options.functions);
            } else {
                this.options.validator = new DefaultResponseValidator();
            }
        }
    }

    /**
     * Adds a result from a `function_call` to the history.
     * @param name Name of the function that was called.
     * @param results Results returned by the function.
     */
    public addFunctionResultToHistory(name: string, results: any): void {
        // Convert content to string
        let content = '';
        if (typeof results === "object") {
            if (typeof results.toISOString == "function") {
                content = results.toISOString();
            } else {
                content = JSON.stringify(results);
            }
        } else if (results !== undefined && results !== null) {
            content = results.toString();
        }

        // Add result to history
        const { memory, history_variable } = this.options;
        const history: Message[] = memory.get(history_variable) ?? [];
        history.push({ role: 'function', name, content });
        if (history.length > this.options.max_history_messages) {
            history.splice(0, history.length - this.options.max_history_messages);
        }
        memory.set(history_variable, history);
    }

    /**
     * Completes a prompt.
     * @remarks
     * The `input` parameter is optional but if passed in, will be assigned to memory using the
     * configured `input_variable`. If it's not passed in an attempt will be made to read it
     * from memory so passing it in or assigning to memory works. In either case, the `input`
     * variable is only used when constructing a user message that, will be added to the
     * conversation history and formatted like `{ role: 'user', content: input }`.
     *
     * It's important to note that if you want the users input sent to the model as part of the
     * prompt, you will need to add a `UserMessage` section to your prompt. The wave does not do
     * anything to modify your prompt, except when performing repairs and those changes are
     * temporary.
     *
     * When the model successfully returns a valid (or repaired) response, a 'user' message (if
     * input was detected) and 'assistant' message will be automatically added to the conversation
     * history. You can disable that behavior by setting `max_history_messages` to `0`.
     *
     * The response returned by `completePrompt()` will be strongly typed by the validator you're
     * using. The `DefaultResponseValidator` returns a `string` and the `JSONResponseValidator`
     * will return either an `object` or if a JSON Schema is provided, an instance of `TContent`.
     * When using a custom validator, the validator is return any type of content it likes.
     *
     * A successful response is indicated by `response.status == 'success'` and the content can be
     * accessed via `response.message.content`.  If a response is invalid it will have a
     * `response.status == 'invalid_response'` and the `response.message` will be a string containing
     * the validator feedback message.  There are other status codes for various errors and in all
     * cases except `success` the `response.message` will be of type `string`.
     * @template TContent Optional. Type of message content returned for a 'success' response. The `response.message.content` field will be of type TContent. Defaults to `any`.
     * @param input Optional. Input text to use for the user message.
     * @returns A strongly typed response object.
     */
    public async completePrompt<TContent = any>(input?: string): Promise<PromptResponse<TContent>> {
        const { prompt, memory, functions, tokenizer, validator, max_repair_attempts, history_variable, input_variable } = this.options;
        let { model } = this.options;

        // Check for OpenAI model being used with a function validator
        if (model instanceof OpenAIModel && validator instanceof FunctionResponseValidator && !model.options.functions) {
            // Create a clone of the model that's configured to use the validators functions
            model = model.clone({ functions: validator.functions })
        }

        // Update/get user input
        if (input_variable) {
            if (typeof input === 'string') {
                memory.set(input_variable, input);
            } else {
                input = memory.has(input_variable) ? memory.get(input_variable) : ''
            }
        } else if (!input) {
            input = '';
        }

        try {
            // Ask client to complete prompt
            this.emit('beforePrompt', memory, functions, tokenizer, prompt);
            const response = await model.completePrompt(memory, functions, tokenizer, prompt);
            this.emit('afterPrompt', memory, functions, tokenizer, prompt, response);
            if (response.status !== 'success') {
                return response;
            }

            // Ensure response is a message
            if (typeof response.message !== 'object') {
                response.message = { role: 'assistant', content: response.message ?? '' };
            }

            // Validate response
            this.emit('beforeValidation', memory, functions, tokenizer, response, max_repair_attempts);
            const validation = await validator.validateResponse(memory, functions, tokenizer, response, max_repair_attempts);
            this.emit('afterValidation', memory, functions, tokenizer, response, max_repair_attempts, validation);
            if (validation.valid) {
                // Update content
                if (validation.hasOwnProperty('value')) {
                    response.message.content = validation.value;
                }

                // Update history and return
                this.addInputToHistory(memory, history_variable, input!);
                this.addResponseToHistory(memory, history_variable, response.message);
                return response;
            }

            // Bail out if we're not repairing
            if (max_repair_attempts <= 0) {
                return response;
            }

            // Fork the conversation history
            const fork = new MemoryFork(memory);

            // Log repair attempts
            if (this.options.logRepairs) {
                console.log(Colorize.title('REPAIRING RESPONSE:'));
                console.log(Colorize.output(response.message.content));
            }

            // Attempt to repair response
            this.emit('beforeRepair', fork, functions, tokenizer, response, max_repair_attempts, validation);
            const repair = await this.repairResponse(fork, functions, tokenizer, response, validation, max_repair_attempts);
            this.emit('afterRepair', fork, functions, tokenizer, response, max_repair_attempts, validation);

            // Log repair success
            if (this.options.logRepairs) {
                if (repair.status === 'success') {
                    console.log(Colorize.success('Response Repaired'));
                } else {
                    console.log(Colorize.error('Response Repair Failed'));
                }
            }

            // Update history with repaired response if successful.
            // - conversation history will be left unchanged if the repair failed.
            // - we never want to save an invalid response to conversation history.
            // - the caller can take further corrective action, including simply re-trying.
            if (repair.status === 'success') {
                this.addInputToHistory(memory, history_variable, input!);
                this.addResponseToHistory(memory, history_variable, repair.message as Message);
            }

            return repair;
        } catch (err: unknown) {
            return {
                status: 'error',
                message: err instanceof Error ? err.message : `${err}`
            };
        }
    }

    /**
     * @private
     */
    private addInputToHistory(memory: PromptMemory, variable: string, input: string): void {
        if (variable && input.length > 0) {
            const history: Message[] = memory.get(variable) ?? [];
            history.push({ role: 'user', content: input });
            if (history.length > this.options.max_history_messages) {
                history.splice(0, history.length - this.options.max_history_messages);
            }
            memory.set(variable, history);
        }
    }

    /**
     * @private
     */
    private addResponseToHistory(memory: PromptMemory, variable: string, message: Message): void {
        if (variable) {
            const history: Message[] = memory.get(variable) ?? [];
            history.push(message);
            if (history.length > this.options.max_history_messages) {
                history.splice(0, history.length - this.options.max_history_messages);
            }
            memory.set(variable, history);
        }
    }

    /**
     * @private
     */
    private async repairResponse(fork: MemoryFork, functions: PromptFunctions, tokenizer: Tokenizer, response: PromptResponse, validation: Validation, remaining_attempts: number): Promise<PromptResponse> {
        const { model, prompt, validator } = this.options;

        // Add response and feedback to repair history
        const feedback = validation.feedback ?? 'The response was invalid. Try another strategy.';
        this.addResponseToHistory(fork, `${this.options.history_variable}-repair`, response.message as Message);
        this.addInputToHistory(fork, `${this.options.history_variable}-repair`, feedback);

        // Append repair history to prompt
        const repairPrompt = new Prompt([
            prompt,
            new ConversationHistory(`${this.options.history_variable}-repair`)
        ]);

        // Log the repair
        if (this.options.logRepairs) {
            console.log(Colorize.value('feedback', feedback));
        }

        // Ask client to complete prompt
        this.emit('beforePrompt', fork, functions, tokenizer, prompt);
        const repairResponse = await model.completePrompt(fork, functions, tokenizer, repairPrompt);
        this.emit('afterPrompt', fork, functions, tokenizer, prompt, repairResponse);
        if (repairResponse.status !== 'success') {
            return repairResponse;
        }

        // Ensure response is a message
        if (typeof repairResponse.message !== 'object') {
            repairResponse.message = { role: 'assistant', content: repairResponse.message ?? '' };
        }

        // Validate response
        this.emit('beforeValidation', fork, functions, tokenizer, repairResponse, remaining_attempts);
        validation = await validator.validateResponse(fork, functions, tokenizer, repairResponse, remaining_attempts);
        this.emit('afterValidation', fork, functions, tokenizer, repairResponse, remaining_attempts, validation);
        if (validation.valid) {
            // Update content
            if (validation.hasOwnProperty('value')) {
                repairResponse.message.content = validation.value;
            }

            return repairResponse;
        }

        // Are we out of attempts?
        remaining_attempts--;
        if (remaining_attempts <= 0) {
            return {
                status: 'invalid_response',
                message: validation.feedback ?? 'The response was invalid. Try another strategy.'
            };
        }

        // Try next attempt
        this.emit('nextRepair', fork, functions, tokenizer, repairResponse, remaining_attempts, validation);
        return await this.repairResponse(fork, functions, tokenizer, repairResponse, validation, remaining_attempts);
    }
}

File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/AI/Clients/LLMClient.cs
using Google.Protobuf.WellKnownTypes;
using Microsoft.Bot.Builder;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.Teams.AI.AI.Models;
using Microsoft.Teams.AI.AI.Prompts;
using Microsoft.Teams.AI.AI.Prompts.Sections;
using Microsoft.Teams.AI.AI.Validators;
using Microsoft.Teams.AI.Application;
using Microsoft.Teams.AI.State;
using static Microsoft.Teams.AI.AI.Models.IPromptCompletionModelEvents;

namespace Microsoft.Teams.AI.AI.Clients
{
    /// <summary>
    /// LLMClient class that's used to complete prompts.
    /// </summary>
    /// <remarks>
    /// Each LLMClient, at a minimum needs to be configured with a <see cref="LLMClientOptions{TContent}.Model"/> and <see cref="LLMClientOptions{TContent}.Template"/>.
    ///
    /// Configuring the LLMClient to use a <see cref="LLMClientOptions{TContent}.Validator"/> is optional but recommended. The primary benefit to
    /// using LLMClient is it's response validation and automatic response repair features. The
    /// validator acts as guard and guarantees that you never get an malformed response back from the
    /// model. At least not without it being flagged as an `invalid_response`.
    ///
    /// Using the <see cref="JsonResponseValidator"/>, for example, guarantees that you only ever get a valid
    /// object back from <see cref="CompletePromptAsync(ITurnContext, IMemory, IPromptFunctions{List{string}}, CancellationToken)"/>. In fact, you'll get back a fully parsed object and any
    /// additional response text from the model will be dropped. If you give the <see cref="JsonResponseValidator"/>
    /// a JSON Schema, you will get back a strongly typed and validated instance of an object in
    /// the returned `response.message.content`.
    ///
    /// When a validator detects a bad response from the model, it gives the model "feedback" as to the
    /// problem it detected with its response and more importantly an instruction that tells the model
    /// how it should repair the problem. This puts the LLMClient into a special repair mode where it first
    /// forks the memory for the conversation and then has a side conversation with the model in an
    /// effort to get it to repair its response. By forking the conversation, this isolates the bad
    /// response and prevents it from contaminating the main conversation history. If the response can
    /// be repaired, the LLMClient will un-fork the memory and use the repaired response in place of the
    /// original bad response. To the model it's as if it never made a mistake which is important for
    /// future turns with the model. If the response can't be repaired, a response status of
    /// `invalid_response` will be returned.
    ///
    /// When using a well designed validator, like the <see cref="JsonResponseValidator"/>, the LLMClient can typically
    /// repair a bad response in a single additional model call. Sometimes it takes a couple of calls
    /// to effect a repair and occasionally it won't be able to repair it at all. If your prompt is
    /// well designed and you only occasionally see failed repair attempts, I'd recommend just calling
    /// the LLMClient a second time. Given the stochastic nature of these models, there's a decent chance
    /// it won't make the same mistake on the second call. A well designed prompt coupled with a well
    /// designed validator should get the reliability of calling these models somewhere close to 99%
    /// reliable.
    ///
    /// This "feedback" technique works with all the GPT-3 generation of models and I've tested it with
    /// `text-davinci-003`, `gpt-3.5-turbo`, and `gpt-4`. There's a good chance it will work with other
    /// open source models like `LLaMA` and Googles `Bard` but I have yet to test it with those models.
    /// </remarks>
    /// <typeparam name="TContent">
    /// Type of message content returned for a 'success' response. The `response.message.content` field will be of type TContent.
    /// </typeparam>
    public class LLMClient<TContent>
    {
        /// <summary>
        /// Options for this LLMClient instance.
        /// </summary>
        public readonly LLMClientOptions<TContent> Options;

        private readonly ILogger _logger;

        private readonly string? _startStreamingMessage;
        private ResponseReceivedHandler? _endStreamHandler;
        private bool? _enableFeedbackLoop;

        /// <summary>
        /// Creates a new `LLMClient` instance.
        /// </summary>
        /// <param name="options">Options for this LLMClient instance.</param>
        /// <param name="loggerFactory">logger</param>
        public LLMClient(LLMClientOptions<TContent> options, ILoggerFactory? loggerFactory = null)
        {
            this.Options = options;
            this._logger = loggerFactory?.CreateLogger(typeof(LLMClient<TContent>)) ?? NullLogger.Instance;

            if (options.LogRepairs && loggerFactory == null)
            {
                throw new ArgumentException($"`{nameof(loggerFactory)}` parameter cannot be null if `LogRepairs` option is set to true");
            }

            this._startStreamingMessage = Options.StartStreamingMessage;
            this._endStreamHandler = Options.EndStreamHandler;
            this._enableFeedbackLoop = Options.EnableFeedbackLoop;
        }

        /// <summary>
        /// Adds a result from a `function_call` to the history.
        /// </summary>
        /// <param name="memory">An interface for accessing state values.</param>
        /// <param name="name">Name of the function that was called.</param>
        /// <param name="results">Results returned by the function.</param>
        public void AddFunctionResultToHistory(IMemory memory, string name, object results)
        {
            List<ChatMessage> history = (List<ChatMessage>?)memory.GetValue(this.Options.HistoryVariable) ?? new() { };

            history.Add(new(ChatRole.Function)
            {
                Name = name,
                Content = results.ToString()
            });

            if (history.Count > this.Options.MaxHistoryMessages)
            {
                history.RemoveAt(0);
            }

            memory.SetValue(this.Options.HistoryVariable, history);
        }

        /// <summary>
        /// Completes a prompt.
        /// </summary>
        /// <remarks>
        /// The `input` parameter is optional but if passed in, will be assigned to memory using the
        /// configured `InputVariable`. If it's not passed in an attempt will be made to read it
        /// from memory so passing it in or assigning to memory works. In either case, the `input`
        /// variable is only used when constructing a user message that, will be added to the
        /// conversation history and formatted like `{ role: 'user', content: input }`.
        ///
        /// It's important to note that if you want the users input sent to the model as part of the
        /// prompt, you will need to add a `UserMessageSection` to your prompt. The LLMClient does not do
        /// anything to modify your prompt, except when performing repairs and those changes are
        /// temporary.
        ///
        /// When the model successfully returns a valid (or repaired) response, a 'user' message (if
        /// input was detected) and 'assistant' message will be automatically added to the conversation
        /// history. You can disable that behavior by setting `MaxHistoryMessages` to `0`.
        ///
        /// The response returned by `CompletePromptAsync()` will be strongly typed by the validator you're
        /// using. The `DefaultResponseValidator` returns a `string` and the `JsonResponseValidator`
        /// will return either an `object` or if a JSON Schema is provided, an instance of `TContent`.
        /// When using a custom validator, the validator is return any type of content it likes.
        ///
        /// A successful response is indicated by `response.status == 'success'` and the content can be
        /// accessed via `response.message.content`.  If a response is invalid it will have a
        /// `response.status == 'invalid_response'` and the `response.message` will be a string containing
        /// the validator feedback message.  There are other status codes for various errors and in all
        /// cases except `success` the `response.message` will be of type `string`.
        /// </remarks>
        /// <param name="context">Current turn context.</param>
        /// <param name="memory">An interface for accessing state values.</param>
        /// <param name="functions">Functions to use when rendering the prompt.</param>
        /// <param name="cancellationToken"></param>
        /// <returns>A `PromptResponse` with the status and message.</returns>
        public async Task<PromptResponse> CompletePromptAsync(
            ITurnContext context,
            IMemory memory,
            IPromptFunctions<List<string>> functions,
            CancellationToken cancellationToken = default
        )
        {
            // Define event handlers
            StreamingResponse? streamer = null;

            BeforeCompletionHandler handleBeforeCompletion = new((object sender, BeforeCompletionEventArgs args) =>
            {
                // Ignore events for other contexts
                if (args.TurnContext != context)
                {
                    return;
                }

                if (args.Streaming)
                {
                    // Attach to any existing streamer
                    // - see tool call note below to understand.
                    streamer = (StreamingResponse?)memory.GetValue("temp.streamer");
                    if (streamer == null)
                    {
                        // Create streamer and send initial message
                        streamer = new StreamingResponse(context);
                        memory.SetValue("temp.streamer", streamer);

                        if (this._enableFeedbackLoop != null)
                        {
                            streamer.EnableFeedbackLoop = this._enableFeedbackLoop;
                        }

                        streamer.EnableGeneratedByAILabel = true;

                        if (!string.IsNullOrEmpty(this._startStreamingMessage))
                        {
                            streamer.QueueInformativeUpdate(this._startStreamingMessage!);
                        }
                    }
                }
            });

            ChunkReceivedHandler handleChunkReceived = new((object sender, ChunkReceivedEventArgs args) =>
            {
                if (args.TurnContext != context || streamer == null)
                {
                    return;
                }


                // Ignore content without text
                // - The chunk is likely for a Tool Call.
                // - See the tool call note below to understand why we're ignoring them.
                if (args.Chunk.delta?.GetContent<string>() == null)
                {
                    return;
                }

                // Send chunk to client
                string text = args.Chunk.delta?.GetContent<string>() ?? "";
                IList<Citation>? citations = args.Chunk.delta?.Context?.Citations ?? null;


                if (text.Length > 0)
                {
                    streamer.QueueTextChunk(text, citations);
                }
            });

            // Subscribe to model events
            if (this.Options.Model is IPromptCompletionStreamingModel)
            {
                IPromptCompletionStreamingModel model = (IPromptCompletionStreamingModel)Options.Model;

                if (model.Events != null)
                {
                    model.Events.BeforeCompletion += handleBeforeCompletion;
                    model.Events.ChunkReceived += handleChunkReceived;

                    if (this._endStreamHandler != null)
                    {
                        model.Events.ResponseReceived += this._endStreamHandler;
                    }
                }
            }

            try
            {
                PromptResponse response = await this.Options.Model.CompletePromptAsync(
                    context,
                    memory,
                    functions,
                    this.Options.Tokenizer,
                    this.Options.Template,
                    cancellationToken
                );

                // Handle streaming responses
                if (streamer != null)
                {
                    // Tool call handling
                    // - We need to keep the streamer around during tool calls so we're just letting them return as normal
                    //   messages minus the message content. The text content is being streamed to the client in chunks.
                    // - When the tool call completes we'll call back into ActionPlanner and end up re-attaching to the
                    //   streamer. This will result in us continuing to stream the response to the client. 
                    if (response.Message?.ActionCalls != null && response.Message.ActionCalls.Count > 0)
                    {
                        // Ensure content is empty for tool calls
                        response.Message.Content = "";
                    }
                    else
                    {
                        if (response.Status == PromptResponseStatus.Success)
                        {
                            // Delete message from response to avoid sending it twice
                            response.Message = null;
                        }

                        // End the stream and remove pointer from memory
                        // - We're not listening for the response received event because we can't await the completion of events.
                        await streamer.EndStream();
                        memory.DeleteValue("temp.streamer");
                    }
                }

                // Get input message/s
                string inputVariable = Options.InputVariable;
                IList<ChatMessage>? inputMsgs = response.Input;
                if (inputMsgs == null)
                {
                    object? content = memory.GetValue(inputVariable);
                    inputMsgs = new List<ChatMessage>() { new ChatMessage(ChatRole.User) };
                    inputMsgs[0].Content = content;
                }

                Validation validation = await this.Options.Validator.ValidateResponseAsync(
                    context,
                    memory,
                    this.Options.Tokenizer,
                    response,
                    this.Options.MaxRepairAttempts,
                    cancellationToken
                );

                if (validation.Valid)
                {
                    if (validation.Value != null && response.Message != null)
                    {
                        response.Message.Content = validation.Value.ToString();
                    }

                    this.AddMessagesToHistory(memory, this.Options.HistoryVariable, inputMsgs);

                    if (response.Message != null)
                    {
                        this.AddMessageToHistory(memory, this.Options.HistoryVariable, response.Message);
                    }

                    return response;
                }

                if (this.Options.MaxRepairAttempts <= 0)
                {
                    return response;
                }

                MemoryFork fork = new(memory);

                if (this.Options.LogRepairs)
                {
                    this._logger.LogInformation("REPAIRING RESPONSE:");
                    this._logger.LogInformation(response.Message?.Content?.ToString() ?? string.Empty);
                }

                PromptResponse repairResponse = await this.RepairResponseAsync(
                    context,
                    fork,
                    functions,
                    response,
                    validation,
                    this.Options.MaxRepairAttempts,
                    cancellationToken
                );

                if (this.Options.LogRepairs)
                {
                    this._logger.LogInformation(repairResponse.Status == PromptResponseStatus.Success ? "Response Repaired" : "Response Repair Failed");
                }

                if (repairResponse.Status == PromptResponseStatus.Success)
                {
                    this.AddMessagesToHistory(memory, this.Options.HistoryVariable, inputMsgs);

                    if (repairResponse.Message != null)
                    {
                        this.AddMessageToHistory(memory, this.Options.HistoryVariable, repairResponse.Message);
                    }
                }

                return repairResponse;
            }
            catch (Exception ex)
            {
                return new()
                {
                    Status = PromptResponseStatus.Error,
                    Error = new(ex.Message ?? string.Empty)
                };
            }
            finally
            {

                // Unsubscribe to model events
                if (this.Options.Model is IPromptCompletionStreamingModel)
                {
                    IPromptCompletionStreamingModel model = (IPromptCompletionStreamingModel)Options.Model;

                    if (model.Events != null)
                    {
                        model.Events.BeforeCompletion -= handleBeforeCompletion;
                        model.Events.ChunkReceived -= handleChunkReceived;
                        if (this._endStreamHandler != null)
                        {
                            model.Events.ResponseReceived -= this._endStreamHandler;
                        }
                    }
                }
            }
        }

        private async Task<PromptResponse> RepairResponseAsync(
            ITurnContext context,
            MemoryFork fork,
            IPromptFunctions<List<string>> functions,
            PromptResponse response,
            Validation validation,
            int remainingAttempts,
            CancellationToken cancellationToken
        )
        {
            string feedback = validation.Feedback ?? "The response was invalid. Try another strategy.";

            this.AddMessageToHistory(fork, $"{this.Options.HistoryVariable}-repair", new(ChatRole.User) { Content = feedback });

            if (response.Message != null)
            {
                this.AddMessageToHistory(fork, $"{this.Options.HistoryVariable}-repair", response.Message);
            }

            PromptTemplate repairTemplate = new(this.Options.Template);
            repairTemplate.Prompt = new(new()
            {
                this.Options.Template.Prompt,
                new ConversationHistorySection($"{this.Options.HistoryVariable}-repair", -1)
            });

            if (this.Options.LogRepairs)
            {
                this._logger.LogInformation(feedback);
            }

            PromptResponse repairResponse = await this.Options.Model.CompletePromptAsync(
                context,
                fork,
                functions,
                this.Options.Tokenizer,
                repairTemplate,
                cancellationToken
            );

            if (repairResponse.Status != PromptResponseStatus.Success)
            {
                return repairResponse;
            }

            Validation repairValidation = await this.Options.Validator.ValidateResponseAsync(
                context,
                fork,
                this.Options.Tokenizer,
                repairResponse,
                remainingAttempts,
                cancellationToken
            );

            if (repairValidation.Valid)
            {
                if (repairValidation.Value != null && repairResponse.Message != null)
                {
                    repairResponse.Message.Content = repairValidation.Value.ToString();
                }

                return repairResponse;
            }

            remainingAttempts--;

            if (remainingAttempts <= 0)
            {
                return new()
                {
                    Status = PromptResponseStatus.InvalidResponse,
                    Error = new($"Reached max model response repair attempts. Last feedback given to model: \"{feedback}\"")
                };
            }

            return await this.RepairResponseAsync(context, fork, functions, repairResponse, repairValidation, remainingAttempts, cancellationToken);
        }

        private void AddMessagesToHistory(IMemory memory, string variable, IEnumerable<ChatMessage> messages)
        {
            if (variable == string.Empty || variable == null)
            {
                return;
            }

            List<ChatMessage> history = (List<ChatMessage>?)memory.GetValue(variable) ?? new();

            history.AddRange(messages);

            if (history.Count > this.Options.MaxHistoryMessages)
            {
                int overflow = history.Count - this.Options.MaxHistoryMessages;
                history.RemoveRange(0, overflow);
            }

            // Remove completed partial action outputs
            while (history.Count > 0 && history[0].Role == ChatRole.Tool)
            {
                history.RemoveAt(0);
            }


            memory.SetValue(variable, history);
        }


        private void AddMessageToHistory(IMemory memory, string variable, ChatMessage message)
        {
            AddMessagesToHistory(memory, variable, new List<ChatMessage>() { message });
        }
    }
}
